pkg/ignition/ignition.go

```diff
@@ -0,0 +1,134 @@
+)
+
+//+kubebuilder:rbac:groups="machineconfiguration.openshift.io",resources=machineconfigs,verbs=list;watch
+
+const (
+	// kubeletSystemdName is the name of the systemd service that the kubelet runs under,
+	// this is used to parse the kubelet args
+	kubeletSystemdName = "kubelet.service"
+	// CloudConfigOption is the kubelet CLI option for the cloud configuration file
+	CloudConfigOption = "cloud-config"
+	// CloudProviderOption is the kubelet CLI option for cloud provider
+	CloudProviderOption = "cloud-provider"
+	// RenderedWorkerPrefix allows identification of the rendered worker MachineConfig, the combination of all worker
+	// MachineConfigs.
+	RenderedWorkerPrefix = "rendered-worker-"
+)
+
+// Ignition is a representation of an Ignition file
+type Ignition struct {
+	config ignCfgTypes.Config
+}
+
+// New returns a new instance of Ignition
+func New(c client.Client) (*Ignition, error) {
+	log := ctrl.Log.WithName("ignition")
+	machineConfigs := &mcfg.MachineConfigList{}
+	err := c.List(context.TODO(), machineConfigs)
+	if err != nil {
+		return nil, err
+	}
+	renderedWorker, err := getLatestRenderedWorker(machineConfigs.Items)
+	if err != nil {
+		return nil, err
+	}
+
+	configuration, report, err := ignCfg.Parse(renderedWorker.Spec.Config.Raw)
+	if err != nil || report.IsFatal() {
+		return nil, errors.Errorf("failed to parse MachineConfig ignition: %v\nReport: %v", err, report)
+	}
+	ign := &Ignition{
+		config: configuration,
+	}
+	log.V(1).Info("parsed", "machineconfig", renderedWorker.GetName(), "ignition version",
+		configuration.Ignition.Version)
+	return ign, nil
+}
+
+// GetKubeletArgs returns a set of arguments for kubelet.exe, as specified in the ignition file
+func (ign *Ignition) GetKubeletArgs() (map[string]string, error) {
+	var kubeletUnit ignCfgTypes.Unit
+	for _, unit := range ign.config.Systemd.Units {
+		if unit.Name == kubeletSystemdName {
+			kubeletUnit = unit
+			break
+		}
+	}
+	if kubeletUnit.Contents == nil {
+		return nil, errors.Errorf("ignition missing kubelet systemd unit file")
+	}
+	argsFromIgnition, err := parseKubeletArgs(*kubeletUnit.Contents)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing kubelet systemd unit args")
+	}
+	return argsFromIgnition, nil
+}
+
+// getLatestRenderedWorker returns the most recently created rendered worker MachineConfig
+func getLatestRenderedWorker(machineConfigs []mcfg.MachineConfig) (*mcfg.MachineConfig, error) {
+	// Grab the latest rendered-worker MachineConfig by sorting the MachineConfig list by the latest creation
+	// timestamp first.
+	sort.Slice(machineConfigs, func(i, j int) bool {
+		iTimestamp := machineConfigs[i].GetCreationTimestamp()
+		jTimestamp := machineConfigs[j].GetCreationTimestamp()
+		return jTimestamp.Before(&iTimestamp)
+	})
+	for _, mc := range machineConfigs {
+		if strings.HasPrefix(mc.Name, RenderedWorkerPrefix) {
+			if len(mc.Spec.Config.Raw) == 0 {
+				continue
+			}
+			return &mc, nil
```

Comment:
Please add an info message here saying which MC we picked. This will help debugging customer issues.